tests/MongoDB.Driver.Tests/ReadPreferenceOnStandaloneTests.cs

```diff
@@ -0,0 +1,86 @@
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+using MongoDB.Driver.Core;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Events;
+using MongoDB.Driver.Core.Misc;
+using MongoDB.Driver.TestHelpers;
+using Xunit;
+
+namespace MongoDB.Driver.Tests
+{
+    public class ReadPreferenceOnStandaloneTests
+    {
+        [Theory]
+        [ParameterAttributeData]
+        public void ReadPreference_should_not_be_sent_to_standalone_server(
+            [Values(false, true)] bool async)
+        {
+            var eventCapturer = new EventCapturer().Capture<CommandStartedEvent>(e => e.CommandName.Equals("find") || e.CommandName.Equals("$query"));
+            using (var client = CreateDisposableClient(eventCapturer, ReadPreference.PrimaryPreferred))
+            {
+                var database = client.GetDatabase(DriverTestConfiguration.DatabaseNamespace.DatabaseName);
+                var collection = database.GetCollection<BsonDocument>(DriverTestConfiguration.CollectionNamespace.CollectionName);
+
+                if (async)
+                {
+                    var _ = collection.FindAsync("{ x : 2 }").GetAwaiter().GetResult();
+                }
+                else
+                {
+                    var _ = collection.FindSync("{ x : 2 }");
+                }
+
+                BsonDocument sentCommand = ((CommandStartedEvent)eventCapturer.Events[0]).Command;;
+                var serverVersion = client.Cluster.Description.Servers[0].Version;
+
+                if (client.Cluster.Description.Type == ClusterType.Standalone)
+                {
+                    sentCommand.Contains("readPreference").Should().BeFalse();
+                }
+                else if (client.Cluster.Description.Type == ClusterType.Sharded &&
+                         serverVersion < Feature.CommandMessage.FirstSupportedVersion)
+                {
+                    if (((CommandStartedEvent) eventCapturer.Events[0]).CommandName.Equals("$query"))
+                    {
+                        sentCommand.Contains("$readPreference").Should().BeTrue();
+                    }
+                    else if (((CommandStartedEvent) eventCapturer.Events[0]).CommandName.Equals("find"))
+                    {
+                        sentCommand.Contains("readPreference").Should().BeTrue();
+                    }
+                }
+                else if (serverVersion >= Feature.CommandMessage.FirstSupportedVersion)
+                {
+                    sentCommand.Contains("$readPreference").Should().BeTrue();
```

Comment:
Were you going to restructure the last two clauses as suggested in a previous comment to avoid duplication of asserts?

https://github.com/ksadoff/mongo-csharp-driver/pull/9#discussion_r310150753

Or even better... if we fix the `$query` issue maybe we don't need different asserts for older and newer servers.

Reply:
I've been investigating the `$query` issue a bit more. It looks like the message comes into `COmmandUsingQueryMessageWireProtocol.cs` looking like `{ "find" : "testcollection", "filter" : { "x" : 2 } }` at [here](https://github.com/ksadoff/mongo-csharp-driver/blob/53220d2268d03e4a2fad2c28aa8f017381dc6095/src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingQueryMessageWireProtocol.cs#L85). It then gets wrapped with the `$query` command [here](https://github.com/ksadoff/mongo-csharp-driver/blob/53220d2268d03e4a2fad2c28aa8f017381dc6095/src/MongoDB.Driver.Core/Core/WireProtocol/CommandUsingQueryMessageWireProtocol.cs#L362) and gets returned that way, since the command has two elements (`$query` and `$readPreference`). This is while I'm testing locally against a 3.2 sharded cluster.

Reply:
The hypothesis all along has been that there is a bug in the transformation of an OP_QUERY based command into a pseudo command for APM purposes.

`$query` is an artifact of OP_QUERY commands. It should not have been present in the pseudo command manufactured for APM. That is a bug in APM.

So the assert in the new test you write is weird. You are asserting against a bug in APM.

@jyemin don't we need to fix APM?

This issue would affect any application that is monitoring events also.

Reply:
Yes, but let's do it as a separate issue/commit rather than complicate this one any further.  We can fix this assertion when we fix that issue.